File: NDPServer/src/main/java/com/slender/config/manager/AuthenticationTokenManager.java
package com.slender.config.manager;

import com.slender.exception.config.ProviderNotFoundException;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.AuthenticationProvider;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.AuthenticationException;

import java.util.HashSet;

public final class AuthenticationTokenManager implements AuthenticationManager {
    private final HashSet<AuthenticationProvider> authenticationProviders=new HashSet<>();
    @Override
    public Authentication authenticate(Authentication authentication) throws AuthenticationException {
        for (AuthenticationProvider provider : authenticationProviders) {
            if (provider.supports(authentication.getClass())) {
                return provider.authenticate(authentication);
            }
        }
        throw new ProviderNotFoundException();
    }

    public AuthenticationTokenManager addProviders(AuthenticationProvider provider){
        this.authenticationProviders.add(provider);
        return this;
    }
}

File: NDPServer/src/main/java/com/slender/config/manager/AuthenticationTokenManager.kt
package com.slender.config.manager

import com.slender.exception.config.ProviderNotFoundException
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.AuthenticationProvider
import org.springframework.security.core.Authentication

class AuthenticationTokenManager : AuthenticationManager {
    private val authenticationProviders = HashSet<AuthenticationProvider>()
    override fun authenticate(authentication: Authentication): Authentication {
        for (provider in authenticationProviders) {
            if (provider.supports(authentication.javaClass)) {
                return provider.authenticate(authentication)
            }
        }
        throw ProviderNotFoundException()
    }

    fun addProviders(provider: AuthenticationProvider): AuthenticationTokenManager {
        this.authenticationProviders.add(provider)
        return this
    }
}